File: src/main/scala/com/vc/hard/L499.scala
package com.vc.hard

object L499 {
  def findShortestWay(arr: Array[Array[Int]], ball: Array[Int], hole: Array[Int]): String = {
    import scala.collection.mutable
    val n = arr.length
    val m = arr(0).length

    def move(x: Int, y: Int, xDir: Int, yDir: Int): (Int, Int, Int) = {
      var xNew = x
      var yNew = y
      var res = 0
      while(
        xNew + xDir >= 0 && xNew + xDir < n &&
          yNew + yDir >= 0 && yNew + yDir < m &&
          arr(xNew + xDir)(yNew + yDir) != 1 &&
          !(xNew + xDir == hole(0) && yNew + yDir == hole(1))
      ) {
        xNew = xNew + xDir
        yNew = yNew + yDir
        res += 1
      }
      if(xNew + xDir == hole(0) && yNew + yDir == hole(1)) (hole(0), hole(1), res + 1)
      else (xNew, yNew, res)
    }

    case class Point(x: Int, y: Int, distance: Int, str: String)
    object PQOrdering extends Ordering[Point] {
      def compare(p1: Point, p2: Point): Int = {
        if(p1.distance == p2.distance) p2.str.compareTo(p1.str)
        else p2.distance - p1.distance
      }
    }
    val dirs = Array(('d', 1, 0), ('l', 0, -1), ('r', 0, 1), ('u', -1, 0))
    val q = new mutable.PriorityQueue[Point]()(PQOrdering)
    val visited = new mutable.HashSet[(Int, Int)]()
    val p = Point(ball(0), ball(1), 0, "")
    q.enqueue(p)
    while(q.nonEmpty) {
      val point = q.dequeue()
      val x = point.x
      val y = point.y
      val res = point.str
      val dist = point.distance
      if(x == hole(0) && y == hole(1)) return res
      visited += ((x, y))
      dirs.foreach(dir => {
        val (xNew, yNew, distance) = move(x, y, dir._2, dir._3)
        val newPoint = Point(xNew, yNew, dist + distance, res + dir._1)
        if(!visited.contains(xNew, yNew)) q.enqueue(newPoint)
      })
    }
    "impossible"
  }
}

File: src/main/java/com/vc/hard/L499.java
package com.vc.hard;

import java.util.*;

class L499 {
    static class Point {
        int x;
        int y;
        int distance;
        String path;

        Point(int x, int y, int distance, String path) {
            this.x = x;
            this.y = y;
            this.distance = distance;
            this.path = path;
        }
    }

    static class PointComparator implements Comparator<Point> {
        public int compare(Point p1, Point p2) {
            int cmp = Integer.valueOf(p1.distance).compareTo(p2.distance);
            if(cmp == 0) return p1.path.compareTo(p2.path);
            return cmp;
        }
    }

    int[][] dirs = {{0, 1},{0, -1},{-1, 0},{1, 0}};
    String[] paths = {"r","l","u","d"};
    private Point move(int[][] maze,
                       Point p, int dirX, int dirY, int n, int m, String path,
                       int targetX, int targetY) {
        int res = 0;
        int x = p.x;
        int y = p.y;

        while(
                x + dirX >= 0 && x + dirX < n &&
                        y + dirY >= 0 && y + dirY < m &&
                        maze[x + dirX][y + dirY] != 1 &&
                        !(x + dirX == targetX && y + dirY == targetY)
        ) {
            x = x + dirX;
            y = y + dirY;
            res++;
        }

        if(x + dirX == targetX && y + dirY == targetY)
            return new Point(x + dirX, y + dirY, p.distance + res + 1, p.path + path);
        else
            return new Point(x, y, p.distance + res, p.path + path);
    }

    public String findShortestWay(int[][] maze, int[] ball, int[] hole) {
        int targetX = hole[0];
        int targetY = hole[1];

        int n = maze.length;
        if(n == 0) return "";
        int m = maze[0].length;

        PriorityQueue<Point> pq = new PriorityQueue<>(new PointComparator());
        pq.offer(new Point(ball[0], ball[1], 0, ""));
        HashSet<String> visited = new HashSet<>();

        while(!pq.isEmpty()) {
            Point e = pq.poll();
            if(e.x == targetX && e.y == targetY) return e.path;
            if(!visited.contains(e.x+"=>"+e.y)) {
                visited.add(e.x+"=>"+e.y);
                //System.out.println("x: "+e.x+" y: "+e.y);
                for(int i = 0; i < dirs.length; i++) {
                    Point newPoint = move(maze, e, dirs[i][0],
                            dirs[i][1], n, m, paths[i], targetX, targetY);
                    if(!visited.contains(newPoint.x+"=>"+newPoint.y)) pq.offer(newPoint);
                }
            }
        }
        return "impossible";
    }
}